metrics/scope.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019 GRAIL, Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package metrics
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"sync/atomic"
+	"unsafe"
+)
+
+// Scope is a collection of metric instances.
+type Scope struct {
+	storage unsafe.Pointer // stores *[]interface{}
+}
+
+// GobEncode implements a custom gob encoder for scopes.
+func (s *Scope) GobEncode() ([]byte, error) {
+	var b bytes.Buffer
+	err := gob.NewEncoder(&b).Encode(s.list())
+	return b.Bytes(), err
+}
+
+// GobDecode implements a custom gob decoder for scopes.
+func (s *Scope) GobDecode(p []byte) error {
+	list := new([]interface{})
+	dec := gob.NewDecoder(bytes.NewReader(p))
+	if err := dec.Decode(list); err != nil {
+		return err
+	}
+	atomic.StorePointer(&s.storage, unsafe.Pointer(list))
+	return nil
+}
+
+// Merge merges instances from Scope u into Scope s.
+func (s *Scope) Merge(u *Scope) {
+	ulist := u.list()
+	if ulist == nil {
+		return
+	}
+	for i, inst := range ulist {
+		if inst == nil {
+			continue
+		}
+		m := metrics[i]
+		m.merge(s.instance(m), inst)
+	}
+}
+
+// Reset resets the scope s to u. It is reset to its initial (zero) state
+// if u is nil.
+func (s *Scope) Reset(u *Scope) {
+	if u == nil {
+		atomic.StorePointer(&s.storage, unsafe.Pointer((uintptr)(0)))
+	} else {
+		atomic.StorePointer(&s.storage, u.storage)
+	}
+}
+
+// instance returns the instance associated with metrics m in the scope s. A new
+// instance is created if none exists yet.
+func (s *Scope) instance(m Metric) interface{} {
+	if inst := s.load(m); inst != nil {
+		return inst
+	}
+	for {
+		ptr := atomic.LoadPointer(&s.storage)
+		var list []interface{}
+		if ptr != unsafe.Pointer((uintptr)(0)) {
+			list = *(*[]interface{})(ptr)
+		}
+		for len(list) <= m.metricID() {
+			list = append(list, nil)
+		}
+		inst := m.newInstance()
+		if inst == nil {
+			panic("metric: metric returned nil instance")
+		}
+		list[m.metricID()] = inst
+		if ok := atomic.CompareAndSwapPointer(&s.storage, ptr, unsafe.Pointer(&list)); ok {
+			return inst
+		}
+	}
```

Comment:
I think this is still racy, as there is nothing preventing parallel access of the aliased backing array of `list`.  I think [this test program](https://play.golang.org/p/ezzj69MpDBe) shows the behavior.  (You must run locally as the  playground does not seem to actual run with parallelism).

Assuming that what I say is true, I think one way to do this without locks is to allocate a slice of `Pointer`s of `len(metrics)` and do the test-and-set on the individual pointers.  That way, there's no need to synchronize access to the backing array.

Reply:
yep, you are right of course. i think a simpler way of dealing with it is to copy the list... this way there can never be any concurrent writes into the list, and concurrent reads are safe...

Reply:
(they should take away my computing license)